### ISSUE TITLE
bc::copy, copy vector data to cube, don't supprot,now!

### ISSUE THREAD

**Issue by xinsuinizhuan** [josephjaspers/blackcat_tensors#65]
vector<vector<double>> matricsInputSeries;
	cube inputs(img_sz, network->m_batch_size, training_sets);
	bc::copy(inputs[i][j].get_stream(),matricsInputSeries[m_batch_size + j].begin(), matricsInputSeries[i * m_batch_size + j].end(), inputs[i][j].cw_begin());

now i can't use bc::copy, copy vector data to cube input!

**Comment by josephjaspers** [comment]
Hi, I've added tests to cover this issue 
https://github.com/josephjaspers/blackcat_tensors/commit/b035502f68c279c85f49d6271dc59e4eaeaf5a88
https://github.com/josephjaspers/blackcat_tensors/commit/6ed9e59db7a4c2d54079f2f33e5e78aacd068b7b

However I was able to successfully compile the above code (Windows VS2019) and have tests run successfully. Without having to change any internal code, so I am unsure why it wouldn't work.

Could you post the error and compilation output?
Thanks!

**Comment by josephjaspers** [comment]
Additionally I haven't yet documented but there is a new class named 

`VecList` which behaves like a regular `bc::Vector` but also has the "push_back" method.
Ergo, you might be able to do something like...

```
bc::VecList<double>  veclist; 
for (int i = 0; i < other_data.size(); ++i) {
     veclist.push_back(other_data[i]); 
}

bc::Cube<double> data(size, batch_sz, other_size); 
data[0][0] = veclist; 
```

**Comment by xinsuinizhuan** [comment]
I meet a strange problem, every functions is ok, when i do not call them. But when i call the function, so many compile errors.
do not call fuction:
![图片](https://user-images.githubusercontent.com/40679769/80552778-44f5a500-89fa-11ea-9b3a-7460c181211e.png)
call function:
![图片](https://user-images.githubusercontent.com/40679769/80552809-60f94680-89fa-11ea-84af-4076dd3d8d81.png)
[error.txt](https://github.com/josephjaspers/blackcat_tensors/files/4549348/error.txt)


[blackcat_tensors_consoledemo.zip](https://github.com/josephjaspers/blackcat_tensors/files/4549329/blackcat_tensors_consoledemo.zip)

**Comment by josephjaspers** [comment]
Hi, I fixed the issue.
I had previously removed utility support (print/copying) from expressions though I have re-added them in the last few commits. 

A while back a re-ordered the template arguments for the Allocator class.
So in your code you have to change

```cpp
	using allocator_type = bc::Allocator<System, value_type>
```
to 
```cpp
        using allocator_type = bc::Allocator<value_type, System>;
```
So if you pull the newest commits and make those changes your code should be able to work. 
(I was able to successfully compile it on my windows VM). 

I added tests to cover this issue. 
Let me know if it works! 

Thanks!

**Comment by xinsuinizhuan** [comment]
thanks, it works. Other problem is that, i predict with the trained data, it seems well, but i predict with the untrained data, it seems not good, i try to decrease the epoch, but it seems not work. I don't know why?

**Comment by josephjaspers** [comment]
Hmmm... I am not sure why, if you send me what you are working on i could take a look?

**Comment by xinsuinizhuan** [comment]
like this:

auto network = bc::nn::neuralnetwork(
		bc::nn::lstm(system_tag, 784/4, 128),
		bc::nn::lstm(system_tag, 128, 64),
		bc::nn::feedforward(system_tag, 64, 10),
		bc::nn::softmax(system_tag, 10),
		bc::nn::logging_output_layer(system_tag, 10, bc::nn::RMSE).skip_every(100)
	);

	bc::print("Neural Network architecture:");
	bc::print(network.get_string_architecture());


	network.set_learning_rate(0.001);
	network.set_batch_size(batch_size);

it break at, network.set_learning_rate(0.001);

[mnist_test.txt](https://github.com/josephjaspers/blackcat_tensors/files/4584436/mnist_test.txt)

**Comment by josephjaspers** [comment]
I will check!

**Comment by josephjaspers** [comment]
Fixed!
https://github.com/josephjaspers/blackcat_tensors/commit/87e97490898caeffd258e032bf7d81973a059be4

Silly bug (infinite recursion), added set_learning_rate to the mnist_recurrent example to catch this bug in the future!
